core/payment/src/api/guard.rs

```diff
@@ -0,0 +1,59 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use tokio::sync::Mutex;
+
+/// Registry of locks for agreements
+pub(super) struct AgreementLock {
+    locks: Mutex<HashMap<String, Arc<Mutex<()>>>>,
+}
+
+impl AgreementLock {
+    /// Construct new instances wrapped in [`Arc`]
+    pub fn arc() -> Arc<Self> {
+        Arc::new(Self::default())
+    }
+
+    /// Take a lock for a given agreement.
+    ///
+    /// The entry in the internal registry will be automatically cleaned up.
+    pub async fn lock(self: &Arc<Self>, agreement: String) -> AgreementLockGuard {
+        let mut map = self.locks.lock().await;
+        let lock = map.entry(agreement).or_default();
+        let guard = Arc::clone(lock).lock_owned().await;
```

Comment:
You acquire lock under previous lock. In case it was already locked, we block execution for all Agreements.